gostatsd.go

```diff
@@ -22,36 +25,126 @@ var (
 	Version string
 )
 
-func main() {
-	runtime.GOMAXPROCS(runtime.NumCPU())
-	s := statsd.NewServer()
-	s.AddFlags(pflag.CommandLine)
-	pflag.Parse()
+const (
+	ParamVerbose    = "verbose"
+	ParamCPUProfile = "cpu-profile"
+	ParamJson       = "json"
+	ParamConfigPath = "config-path"
+	ParamVersion    = "version"
+)
 
-	if s.CPUProfile != "" {
-		f, err := os.Create(s.CPUProfile)
+const EnvPrefix = "GSD" //Go Stats D
+
+func main() {
+	err, v, version := setConfiguration()
+	if err != nil {
+		log.Fatalf("Error while parsing configuration: %v\n", err)
+	}
+	if version {
+		fmt.Printf("Version: %s - Commit: %s - Date: %s\n", Version, GitCommit, BuildDate)
+		return
+	}
+	exitCode := 0
+	defer func(e *int) {
+		//pprof.Lookup("goroutine").WriteTo(os.Stdout, 1)
+		if *e != 0 {
+			os.Exit(*e)
+		}
+	}(&exitCode)
+	CPUProfile := v.GetString(ParamCPUProfile)
+	if CPUProfile != "" {
+		f, err := os.Create(CPUProfile)
 		if err != nil {
-			log.Fatalf("%v", err)
+			log.Fatalf("Failed to open profile file: %v", err)
 		}
+		defer func() {
+			if err := f.Close(); err != nil {
+				log.Errorf("Failed to close profile file: %v", err)
+			}
+		}()
 		pprof.StartCPUProfile(f)
 		defer pprof.StopCPUProfile()
+	}
 
-		c := make(chan os.Signal, 1)
-		signal.Notify(c, os.Interrupt)
-		go func() {
-			for range c {
-				pprof.StopCPUProfile()
-				os.Exit(0)
-			}
-		}()
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+	cancelOnInterrupt(ctx, cancelFunc)
+
+	log.Info("Starting server")
+	s := statsd.Server{
+		Backends:         v.GetStringSlice(statsd.ParamBackends),
```

Comment:
Just found out that this is actually not working because of https://github.com/spf13/viper/issues/112 :( I guess I have to fix it first - it's been open for 7 months already.